### PR TITLE
make GaloisField a subtype of Field

### PR DESCRIPTION
**Pull request by taikiy**
Commandeering @andyleiserson 's #545 plus a bit more refactoring/renaming/cleanup. Thanks Andy!

### CODE REVIEW

File: src/ff/field.rs [hunk 1]
```diff
@@ -1,24 +1,18 @@
-use crate::secret_sharing::SharedValue;
-use generic_array::ArrayLength;
+use crate::secret_sharing::{Block, SharedValue};
 use std::fmt::Debug;
+use typenum::{U1, U4};
 
-// Trait for primitive integer types used to represent the underlying type for field values
-pub trait Int: Sized + Copy + Debug {
```
Comment (taikiy) [inline src/ff/field.rs:6]: This is replaced by `Block` in secret_sharing/mod.rs

File: src/ff/field.rs [hunk 2]
```diff
@@ -1,24 +1,18 @@
...
+    type Size = U4;
+    const VALID_BIT_LENGTH: u32 = u32::BITS;
 }
 
-impl Int for u32 {
-    const BITS: u32 = u32::BITS;
-}
-
-pub trait Field: SharedValue + From<u128> + Into<Self::Integer> {
-    type Integer: Int;
```
Comment (taikiy) [inline src/ff/field.rs:19]: This is replaced by `type Storage: Block` in `SharedValue`.

File: src/ff/field.rs [hunk 3]
```diff
@@ -1,24 +1,18 @@
...
-    type Size: ArrayLength<u8>;
```
Comment (taikiy) [inline src/ff/field.rs:20]: This is moved to `Block` trait.

File: src/ff/galois_field.rs [hunk 4]
```diff
@@ -1,16 +1,48 @@
...
+impl Block for U8_4 {
+    type Size = U4;
+    const VALID_BIT_LENGTH: u32 = 32;
+}
+
+impl Block for U8_5 {
+    type Size = U5;
+    const VALID_BIT_LENGTH: u32 = 40;
+}
+
```
Comment (taikiy) [inline src/ff/galois_field.rs:45]: These are equivalent to former `Int` trait implementations, i.e., storage types.

File: src/ff/galois_field.rs [hunk 5]
```diff
@@ -19,7 +51,7 @@ fn assert_copy<C: Copy>(c: C) -> C {
 }
 
 macro_rules! bit_array_impl {
-    ( $modname:ident, $name:ident, $store:ty, $bits:expr, $arraylen:ty, $polynomial:expr ) => {
```
Comment (taikiy) [inline src/ff/galois_field.rs:22]: Now `$bits` and `$arraylen` are defined in `Block` trait and passed to this macro as `$store`.

File: src/ff/galois_field.rs [hunk 6]
```diff
@@ -47,6 +88,13 @@ macro_rules! bit_array_impl {
                 }
             }
 
+            // TODO (taikiy): Remove this infallible conversion and bring back to TryFrom
+            impl From<u128> for $name {
+                fn from(v: u128) -> Self {
+                    Self::truncate_from(v)
+                }
+            }
```
Comment (taikiy) [inline src/ff/galois_field.rs:96]: GaloisField (or BitArray) used to implement a fallible conversion of TryFrom. By making the `GaloisField` a subtype of `Field`, we need to exclusively choose either `From` or `TryFrom`. I agree with Martin that we should implement TryFrom to catch any unexpected lossy conversions, but there are literary hundreds of references to `F::from()`, so I'm going to make a separate change to remove `From` from `Field` and replace it by `TryFrom`. For now, `GaloisField` needs to follow the trait bound and implement `From` but this is temporal.

File: src/ff/galois_field.rs [hunk 7]
```diff
@@ -201,22 +249,9 @@ macro_rules! bit_array_impl {
...
-                fn try_from(v: u128) -> Result<Self, Self::Error> {
-                    if u128::BITS - v.leading_zeros() <= Self::BITS {
-                        Ok(Self::truncate_from(v))
-                    } else {
-                        Err(format!(
-                            "Bit array size {} is too small to hold the value {}.",
-                            Self::BITS,
-                            v
-                        ))
-                    }
```
Comment (taikiy) [inline src/ff/galois_field.rs:219]: As commented above, this is removed to resolve a conflict, but I will bring this back in the next PR.

File: src/ff/galois_field.rs [hunk 8]
```diff
@@ -422,8 +458,7 @@ bit_array_impl!(
     bit_array_40,
     Gf40Bit,
     U8_5,
-    40,
-    U5,
+    bitarr!(const u8, Lsb0; 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
```
Comment (taikiy) [inline src/ff/galois_field.rs:461]: Field::ONE implementation

File: src/ff/mod.rs [hunk 9]
```diff
@@ -66,23 +66,6 @@ where
 {
 }
 
-/// Trait for data types storing arbitrary number of bits.
-// TODO: Implement `Message`
-pub trait GaloisField:
```
Comment (taikiy) [inline src/ff/mod.rs:71]: This is moved inside galois_field.rs file.

File: src/ff/prime_field.rs [hunk 10]
```diff
@@ -25,41 +28,40 @@ impl<F: PrimeField> Serializable for F {
 }
 
 macro_rules! field_impl {
-    ( $field:ident, $int:ty, $prime:expr, $arraylen:ty ) => {
```
Comment (taikiy) [inline src/ff/prime_field.rs:28]: Same as the `GaloisField` macro. `$arraylen` is now defined in `Block` so no need to pass it here.

**Review by benjaminsavage — APPROVED** [review]
Lgtm!

**Pull request merged**